rule excavator_install:
    # Preprocessing to do a realignment of the reads with GATK
    input:
        sources="%s%sEXCAVATOR_Package_v2.2.tgz" % (config['dirs']['prefix'], config['dirs']['references'])
        #exometrack=lambda wildcards: "%s%s%s" % (config['dirs']['prefix'], config['dirs']['references'], get_reference_exometrack(wildcards.entity, SAMPLESHEETS, config)),
        #references=lambda wildcards: ["%s%s%s%s" % (wildcards.prefix, config['dirs']['references'], get_reference_genome(wildcards.entity, SAMPLESHEETS, config)['file'], ending) for ending in config['endings_bwa']],
        #bam="{prefix}%s%s/{entity}.reCal.reAl.nodup.srt.bam" % (config['dirs']['intermediate'], config['stepnames']['gatk_PrintReads'])
    output:
        "{prefix}%s%s/res.txt" % (config['dirs']['intermediate'], config['stepnames']['excavator_install']),
    log:
        "{prefix}%s%s/res.log" % (config['dirs']['logs'], config['stepnames']['excavator_install'])
    benchmark:
        "{prefix}%s%s/res.benchmark" % (config['dirs']['benchmarks'], config['stepnames']['excavator_install'])
    conda:
        "envs/spike_excavator.yaml"
    threads:
        1
    shell:
        "cwd=`pwd` 2> {log}"
        " && if [ ! -d $CONDA_PREFIX/src/EXCAVATOR/ ]; then"
        " mkdir -p $CONDA_PREFIX/src 2>> {log}"
        " && tar xzvf {input.sources} EXCAVATOR_Package_v2.2/ -C $CONDA_PREFIX/src/ 2>> {log}"
        " && cd $CONDA_PREFIX/src/EXCAVATOR/lib/F77 2>> {log}"
        " && R CMD SHLIB F4R.f 2>> {log}"
        " && R CMD SHLIB FastJointSLMLibraryI.f 2>> {log}"
        " && cd $cwd"
        " fi;"
        " && pwd > {output}"
        " && echo 'fake' >> {output}"
